### PR TITLE
Periodic Docker images build

### PR DESCRIPTION
**Pull request by jtpio**
Related to https://github.com/voila-gallery/gallery/issues/10 and https://github.com/voila-gallery/gallery/issues/18#issuecomment-501742467.

For now this can be tested by doing a [dev install of TLJH](https://tljh.jupyter.org/en/latest/contributing/dev-setup.html) and mounting / installing the plugin in the docker container.

### TODO

- [x] Add script to install and enable the systemd units: `python -m tljh_voila_gallery.install_build_units`
- [x] Make `build_images` build from `gallery.yaml` on the master branch (and default to the package resources)
- [x] ~Fetch new `gallery.yaml` (git pull, `requests.get`...) and trigger `build_images` when the new version has been fetched~
- [x] ~Should the systemd units be installed automatically on plugin install? Could be an issue when developing locally.~ Or run `python -m tljh_voila_gallery.install_build_units` after installing TLJH and the plugin
- [x] ~Fix `WorkingDirectory` for the systemd unit?~ -> remove
- [x] Update build interval: set to 12h
- [x] Rename `install_build_units`
- [x] Prune stopped containers and images before building

### CODE REVIEW

**Comment by jtpio** [comment]
This is the current state:

1. Install TLJH with the `tljh-voila-gallery` plugin
2. Run `sudo /opt/tljh/hub/bin/python3 -m tljh_voila_gallery.install_builder_units` to install the systemd unit files
3. The systemd timer will trigger the service every 12 hours
4. The service runs the command: `python3 -m tljh_voila_gallery.build_images`
5. The  `build_images` script does the following:
    - Prune stopped containers
    - Prune dangling images
    - Build new images from the `gallery.yaml` hosted at https://raw.githubusercontent.com/voila-gallery/gallery/master/tljh-voila-gallery/tljh_voila_gallery/gallery.yaml

Also installing the systemd units automatically (to be triggered on plugin install) is probably not something we want by default. What do you think @yuvipanda?

**Comment by yuvipanda** [comment]
I think we should install it by default. You can put it in https://github.com/voila-gallery/gallery/blob/master/tljh-voila-gallery/tljh_voila_gallery/__init__.py#L61, which is executed only when you install TLJH on a machine by running the bootstrapper. Nothing actually happens if you just install the tljh package, so this won't affect local development.

Are rebuilds no-ops if the image is already cached? If so I'd say we should do this every 30min or so instead of 12h. If not, we should make it so and then do it every 30min :D Doesn't need to block this PR though

**Comment by jtpio** [comment]
> You can put it in https://github.com/voila-gallery/gallery/blob/master/tljh-voila-gallery/tljh_voila_gallery/__init__.py#L61, which is executed only when you install TLJH on a machine by running the bootstrapper. 

Right. Do you think it would also make sense to add a new  `tljh_post_install` hook to TLJH specifically for post install scripts? Even though it's true that the `tljh_config_post_install` or `tljh_custom_jupyterhub_config` hooks can execute arbitrary python code, that could make the intent clearer.

> Are rebuilds no-ops if the image is already cached?

I don't think that's the case for now.

**Comment by yuvipanda** [comment]
> Do you think it would also make sense to add a new `tljh_post_install` hook to TLJH specifically for post install scripts?

Absolutely! Would love a PR!

File: tljh-voila-gallery/tljh_voila_gallery/__init__.py [hunk 1]
```diff
@@ -59,6 +53,7 @@ def login_url(self, base_url):
 
 @hookimpl
 def tljh_custom_jupyterhub_config(c):
+    ensure_builder_units()
```
Comment (yuvipanda) [inline tljh-voila-gallery/tljh_voila_gallery/__init__.py:56]: I just realized this is the wrong place for this. I meant https://github.com/jupyterhub/the-littlest-jupyterhub/blob/master/tljh/hooks.py#L52, which we don't use right now. Should move to a more explicit post-install hook though.

File: tljh-voila-gallery/tljh_voila_gallery/build_images.py [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
+import docker
+import requests
 from ruamel.yaml import YAML
 
 logging.basicConfig(level=logging.INFO)
 
 yaml = YAML()
 
-GALLERY_PATH = 'gallery.yaml'  # Relative to module root
+GALLERY_URL = 'https://raw.githubusercontent.com/voila-gallery/gallery/master/tljh-voila-gallery/tljh_voila_gallery/gallery.yaml'
```
Comment (yuvipanda) [inline tljh-voila-gallery/tljh_voila_gallery/build_images.py]: How would you test this locally? What happens when the format of the yaml file changes but not the code? IMO we should (for now) still load it from the local package, and have some method of autodeployment instead.

File: tljh-voila-gallery/tljh_voila_gallery/__init__.py (hunk 1 above)
Comment (jtpio) [reply tljh-voila-gallery/tljh_voila_gallery/__init__.py:56]: Agreed. The best is indeed to have a dedicated post install hook (will submit a PR for that soon!).

File: tljh-voila-gallery/tljh_voila_gallery/build_images.py (hunk 2 above)
Comment (jtpio) [reply tljh-voila-gallery/tljh_voila_gallery/build_images.py]: Good catch, we should of course support the local gallery.yaml. The same gallery file is also used on the front page so when a new example is added it should also be reflected there.

> What happens when the format of the yaml file changes but not the code? IMO we should (for now) still load it from the local package, and have some method of autodeployment instead.

Maybe one way to ensure consistency between the code and the gallery.yaml would be to reinstall / upgrade the plugin? Followed by the build of the docker images.

Comment (jtpio) [reply tljh-voila-gallery/tljh_voila_gallery/build_images.py]: Rebased to revert back to reading from the local package.

File: tljh-voila-gallery/tljh_voila_gallery/__init__.py (hunk 1 above)
Comment (jtpio) [reply tljh-voila-gallery/tljh_voila_gallery/__init__.py:56]: PR to add the `tljh_post_install` hook to TLJH: https://github.com/jupyterhub/the-littlest-jupyterhub/pull/389

**Comment by jtpio** [comment]
So to trigger a redeployment of the examples, keeping the timer and systemd unit approach, we could think of:

- Re-running the same command to upgrade TLJH with the plugin. Although it doesn't look like TLJH support upgrading plugins? (it does for TLJH itself by passing the `--upgrade` flag)
```
curl https://raw.githubusercontent.com/jupyterhub/the-littlest-jupyterhub/master/bootstrap/bootstrap.py \
 | sudo python3 - \
   --plugin git+https://github.com/voila-gallery/gallery@master#"egg=tljh-voila-gallery&subdirectory=tljh-voila-gallery"
```

- Reinstalling (and updating) the `tljh_voila_gallery` package directly: 
```
sudo /opt/tljh/hub/bin/python3 -m pip install -U git+https://github.com/voila-gallery/gallery@master#"egg=tljh-voila-gallery&subdirectory=tljh-voila-gallery"
```

Both ways would still ensure that `gallery.yaml` and the code stay in sync.

**Comment by yuvipanda** [comment]
@jtpio I currently like approach 2 - reinstalling the tljh_voila_gallery package. We can do that in a different PR though.

File: tljh-voila-gallery/tljh_voila_gallery/systemd-units/tljh-voila-gallery-builder.timer [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description="Build the Docker images for the Voila Gallery Examples"
+
+[Timer]
+OnBootSec=1min
+OnUnitActiveSec=12h
```
Comment (yuvipanda) [inline tljh-voila-gallery/tljh_voila_gallery/systemd-units/tljh-voila-gallery-builder.timer]: Let's bring this down to 2h maybe? 12h seems like a long time :D

Comment (jtpio) [reply tljh-voila-gallery/tljh_voila_gallery/systemd-units/tljh-voila-gallery-builder.timer]: ah good catch, I thought it was already lowered. Should be fixed now.

**Comment by jtpio** [comment]
> We can do that in a different PR though.

Agree, it can be in a follow-up PR.

**Comment by yuvipanda** [comment]
Thanks, @jtpio!

**Comment by jtpio** [comment]
Thanks for reviewing @yuvipanda!

**Pull request merged**